c10/util/ArrayRef.h

```diff
@@ -103,6 +180,48 @@ class ArrayRef final {
     debugCheckNullptrInvariant();
   }
 
+
+  template <size_t N>
+  ArrayRef(const SmallVector<int64_t, N>& Vec):
+    Data(reinterpret_cast<const c10::SymbolicOrConcreteInt*>(Vec.data())), Length(Vec.size()) {
+    debugCheckNullptrInvariant();
+  }
+
+  template <typename U,
+  typename = std::enable_if_t<std::is_constructible<T, U>::value>>
+  ArrayRef(const ArrayRef<U> ar):
+    Data(reinterpret_cast<const T*>(ar.data())), Length(ar.size()) {}
+
+  //template <typename U, typename T1 = T, size_t N,
+  template <typename T1 = T, size_t N,
+  typename = std::enable_if_t<std::is_same<SymbolicOrConcreteInt, T1>::value>>
+  //typename = std::enable_if_t<std::is_constructible<T1, U>::value>>
+  ArrayRef(const SmallVector<int64_t, N>& Vec):
+    Data(reinterpret_cast<const c10::SymbolicOrConcreteInt*>(Vec.data())), Length(Vec.size()) {
+    debugCheckNullptrInvariant();
+  }
+
+  //template <typename U, typename T1 = T, size_t N,
+  template <typename T1 = T, typename T2, typename U,
+  typename = std::enable_if_t<std::is_same<SymbolicOrConcreteInt, T1>::value && 
+    std::is_same<int64_t, T2>::value>>
+  //typename = std::enable_if_t<std::is_constructible<T1, U>::value>>
+  ArrayRef(const SmallVectorTemplateCommon<T2, U>& Vec):
+    Data(reinterpret_cast<const c10::SymbolicOrConcreteInt*>(Vec.data())), Length(Vec.size()) {
+    debugCheckNullptrInvariant();
+  }
+
+
+  template <typename T1 = T, typename T2,
+  typename = std::enable_if_t<std::is_same<SymbolicOrConcreteInt, T1>::value && 
+    std::is_same<int64_t, T2>::value>>
+  ArrayRef(const std::vector<T2>& Vec):
+    Data(reinterpret_cast<const c10::SymbolicOrConcreteInt*>(Vec.data())), Length(Vec.size()) {
+    debugCheckNullptrInvariant();
+  }
```

Comment:
From an engineering perspective, it's a bit bad that we have to explicitly add these new constructors to ArrayRef to get the implicit conversions to work out. I probably would drop them, if it doesn't matter too much (not sure.)